near-plugins-derive/src/upgradable.rs

```diff
@@ -21,20 +25,56 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
         .code_storage_key
         .unwrap_or_else(|| "__CODE__".to_string());
 
+    let staging_timestamp_storage_key = opts
+        .staging_timestamp_storage_key
+        .unwrap_or_else(|| "__TIMESTAMP__".to_string());
+
+    let staging_duration_storage_key = opts
+        .staging_duration_storage_key
+        .unwrap_or_else(|| "__DURATION__".to_string());
+
+    let update_staging_duration_storage_key = opts
+        .update_staging_duration_storage_key
+        .unwrap_or_else(|| "__UPDATE_DURATION__".to_string());
+
+    let update_staging_duration_timestamp_storage_key = opts
+        .update_staging_duration_timestamp_storage_key
+        .unwrap_or_else(|| "__UPDATE_DURATION_TIMESTAMP__".to_string());
+
     let output = quote! {
         #[near_bindgen]
         impl Upgradable for #ident {
-            fn up_storage_key(&self) -> Vec<u8>{
+            fn up_storage_key(&self) -> Vec<u8> {
                 (#code_storage_key).as_bytes().to_vec()
             }
 
+            fn up_staging_timestamp_storage_key(&self) -> Vec<u8> {
+                (#staging_timestamp_storage_key).as_bytes().to_vec()
+            }
+
+            fn up_staging_duration_storage_key(&self) -> Vec<u8> {
+                (#staging_duration_storage_key).as_bytes().to_vec()
+            }
+
+            fn up_update_staging_duration_storage_key(&self) -> Vec<u8> {
+                (#update_staging_duration_storage_key).as_bytes().to_vec()
+            }
+
+            fn up_update_staging_duration_timestamp_storage_key(&self) -> Vec<u8> {
+                (#update_staging_duration_timestamp_storage_key).as_bytes().to_vec()
+            }
+
```

Comment:
Could these values be collected in a struct? Maybe something like:

```rust
#[derive(BorshSerialize, BorshDeserialize)]
struct Delay {
    /// After staging code at least this amount of time must pass before the
    /// code can be deployed.
    delay: Option<Duration>,
    /// If code is staged, it cannot be deployed before this date.
    earliest_deployment: Option<near_sdk::Timestamp>,
}

// Borsh methods for (de)serialization from/to storage:
// * https://docs.rs/borsh/0.9.3/borsh/ser/trait.BorshSerialize.html#method.try_to_vec
// * https://docs.rs/borsh/0.9.3/borsh/de/trait.BorshDeserialize.html#method.try_from_slice
```

Possible benefits:
- Reduce the number of methods in the `Upgradable` trait: there's only one storage key and one getter to get the whole struct might be enough.
- Less (de)serialization in the implementation of `Upgradable`, which makes it easier to follow what's happening.

Reply:
This is a double-edged sword, logically those values aren't used together so extra serialization will be done, but on the other hand, I agree that the code will look more elegant.

@birchmd Based on your experience of managing the storage in the Aurora engine, do you think that grouping keys into structures is a good idea? Were there any problems with migration due to this?

Reply:
The engine keeps values under separate keys for the most part (as opposed to multiple values in a struct); in fact there are some examples where it probably would have been better to group the values under a single key because they are logically correlated (for example account nonce and balance are almost always both needed together).

In general I think it is more important for the storage layout to match the logical usage of the data (data that is accessed together is stored together), as opposed to grouping everything in one value for code reasons. That said, maybe the code could be made to look a little nicer while keeping the values under separate keys. In the engine we use an enum to separate out the different storage keys we have, but then only a single method for accessing the storage (this method takes the enum as an argument). You can also have specialized read and write methods for common types to hide the serialization logic at the main call-site. In this case this might look something like

```rust
enum UpgradableStorageKey {
    StagingTimestamp,
    StagingDuration,
    UpdateStagingTimestamp,
    UpdateStagingDuration,
}

impl AsRef<[u8]> for UpgradableStorageKey { ... }

fn up_get_update_staging_duration(&self) -> Option<near_sdk::Duration> {
    read_duration(UpgradableStorageKey::UpdateStagingDuration)
}

fn up_get_update_staging_duration_timestamp(&self) -> Option<near_sdk::Timestamp> {
    read_timestamp(&UpgradableStorageKey::UpdateStagingTimestamp)
}

fn read_duration(key: &UpgradableStorageKey) -> Option<near_sdk::Duration> {
    near_sdk::env::storage_read(key.as_ref()).map(...)
}
```

What would you think about that @mooori ?

Reply:
That’s a neat way to hide interaction with storage while at the same time avoiding extra de-serialization!

Since near-plugins already depends on near-sdk, we could derive [BorshStorageKey](https://docs.rs/near-sdk/latest/near_sdk/derive.BorshStorageKey.html) for the enum.

Either way, we should be careful not to use storage keys that conflict which storage keys defined by the contract in which the plugin is included. Here’s how we did that in the Acl plugin:

https://github.com/aurora-is-near/near-plugins/blob/7454bbcfcdb50addee5fefe3742879d1fac75c81/near-plugins-derive/src/access_controllable.rs#L83-L99

Reply:
@mooori I've refactored the storage management for the upgradable plugin here a0d0e494f2203ffacee7bd422187d3b6cca96322